services.go

```
@@ -0,0 +1,162 @@
+	services := make([]*Service, len(ss))
+	for _, svc := range ss {
+		services = append(services, svc)
+	}
+
+	payload := struct {
+		Data []*Service `json:"data"`
+	}{services}
+	return json.Marshal(payload)
+}
+
+func (s *Service) logger() *log.Entry {
+	return log.WithFields(log.Fields{
+		"name":         s.Name,
+		"version":      s.cmd.Version,
+		"status":       s.cmd.Status(),
+		"repo_version": s.repo.CurrentHead(),
+	})
+}
+
+// Update search the given command in the map and runs its
+// Update function. If the command is not found, an error is
+// returned.
+func (ss *ServiceStore) Update(name string) error {
+	svc, ok := (*ss)[name]
+
+	if !ok {
+		return fmt.Errorf("services: service with name %q not found", name)
+	}
+
+	cmd := svc.cmd
+
+	svc.logger().Info("Running update")
+	cmd.SetStatus(command.StatusUpdating)
+	updated, err := cmd.Update()
+	if err != nil {
+		svc.cmd.SetStatus(command.StatusRunning)
+		return err
+	}
+
+	if !updated {
+		return nil
+	}
+
+	if err := cmd.Stop(); err != nil {
+		return err
+	}
+
+	updater := cmd.Updater
+	cmd = command.NewCommand(cmd.Log, svc.Name, cmd.Version, cmd.ExecName, cmd.ExecArgs...)
+	cmd.Updater = updater
+	(*ss)[name] = newService(svc.Name, cmd, svc.repo)
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Add builds and add the service to the list
+func (ss *ServiceStore) Add(name string, c *command.Command, r *git.Repo) {
+	s := newService(name, c, r)
+	(*ss)[s.Name] = s
+}
+
+// StartService starts a specific service inside the store.
+// If the service is not found in the list, it returns an error.
+func (ss *ServiceStore) StartService(name string) error {
+	svc, ok := (*ss)[name]
+	if !ok {
+		return fmt.Errorf("services: service %q not found for starting", name)
+	}
+
+	return svc.cmd.Start()
+}
+
+// Start starts all the commands inside the command list by
+// looping and calling each command Start function.
+func (ss *ServiceStore) Start() error {
```

In a case of error what happens? Should we try at least to stop started services?

---

Also maybe is a good idea to known if a service is already running to not start a started command.
Or is already implemented that way?

---

I'm not sure, it can get a little messy, for example, if the application already started and this functions got called when the process was updating, we should not stop other processes, but if the app is just booting, we should stop other process. Maybe that's the approach we need to follow

---

To answer your second question, we are not checking if the process was already started, but if you tried tu run the process, it will throw an error either way ... we should add that validation.

---

related to the first point, maybe we should ignore errors and try to start as many processes as possible and just marked failed ones with a status `StatusError`. But is seems kind of arbitrary start all the process until one processes fail.
I think we should define a predictable behaviour.

---

Maybe we can add one more method `MustStart` that panics in case of a failure and `Start` don't return an error at all.

---

I agree, but if one process fails to start when the operator is booting up, maybe is better approach to no start at all, since is crucial to have al processes running. On the other hand, if a process, after updating, fails to start, is not THAT critical, but it still critical (if you know what I mean), maybe, in that case, we can try to perform a rollback or are least try to fix it

---

Definetly, I'll add that method, that'd resolve our booting problem

---

But if it panics, we should recover from that panic, and stop the already running processes and continue with the gracefully shutdown

---

Ahh, that's true, then return an error and try to stop all process should be a valid approach.

---

I need to check what happens when I try to stop a non-running process. Maybe we can just omit that error (that's how is programmed right now)